# Exploring eBay Car Sales Data
## A Guided Project on DataQuest


#### The data set being analyzed is a sample of 50,000 ads posted on eBay Kleinanzeigen, a classifieds section of the German eBay website. The purpose of this project is to display data cleaning skills and to perform simple analysis. 

In [251]:
import pandas as pd
import numpy as np

# Read file into pandas
autos = pd.read_csv('autos.csv', encoding='Latin-1')

In [252]:
autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
5      2016-03-21 13:47:45  Chrysler_Grand_Voyager_2.8_CRD_Aut.Limited_Sto...   
6      2016-03-20 17:55:21  VW_Golf_III_GT_Special_Electronic_Green_Metall...   
7      2016-03-16 18:55:19                               Golf_IV_1.9_TDI_90PS   
8      2016-03-22 16:51:34                                         Seat_Arosa   
9      2016-03-16 13:47:02          Renault_Megane_Scenic_1.6e_RT_Klimaanlage   
10     2016-03-15 01:41:36                       VW_Golf_Tuning_in_siber/grau   
11     2016-03-16 18:45:34                         Mercedes_A140_Motorschaden   
12     2016-03-31 19:48:22  Smart_smart_fortwo_coupe_softouch_pure_MHD_Pan...   
13     2016-03-23 10:48:32                                 Audi_A3_1.6_tuning   
14     2016-03-23 11:50:46  Renault_Clio_3__Dynamique_1.2__16_V;_viele_Ver...   
15     2016-04-01 12:06:20        Corvette_C3_Coupe_T_Top_Crossfire_Injection   
16     2016-03-16 14:59:02                                Opel_Vectra_B_Kombi   
17     2016-03-29 11:46:22                          Volkswagen_Scirocco_2_G60   
18     2016-03-26 19:57:44               Verkaufen_mein_bmw_e36_320_i_touring   
19     2016-03-17 13:36:21        mazda_tribute_2.0_mit_gas_und_tuev_neu_2018   
20     2016-03-05 19:57:31  Audi_A4_Avant_1.9_TDI_*6_Gang*AHK*Klimatronik*...   
21     2016-03-06 19:07:10                      Porsche_911_Carrera_4S_Cabrio   
22     2016-03-28 20:50:54                               MINI_Cooper_S_Cabrio   
23     2016-03-10 19:55:34   Peugeot_Boxer_2_2_HDi_120_Ps_9_Sitzer_inkl_Klima   
24     2016-04-03 11:57:02                         BMW_535i_xDrive_Sport_Aut.   
25     2016-03-21 21:56:18  Ford_escort_kombi_an_bastler_mit_ghia_ausstattung   
26     2016-04-03 22:46:28                                Volkswagen_Polo_Fox   
27     2016-03-27 18:45:01             Hat_einer_Ahnung_mit_Ford_Galaxy_HILFE   
28     2016-03-19 21:56:19                                      MINI_Cooper_D   
29     2016-04-02 12:45:44  Mercedes_Benz_E_320_T_CDI_Avantgarde_DPF7_Sitz...   
...                    ...                                                ...   
49970  2016-03-21 22:47:37  c4_Grand_Picasso_mit_Automatik_Leder_Navi_Temp...   
49971  2016-03-29 14:54:12                                         W.Lupo_1.0   
49972  2016-03-26 22:25:23          Mercedes_Benz_Vito_115_CDI_Extralang_Aut.   
49973  2016-03-27 05:32:39                   Mercedes_Benz_SLK_200_Kompressor   
49974  2016-03-20 10:52:31  Golf_1_Cabrio_Tuev_Neu_viele_Extras_alles_eing...   
49975  2016-03-27 20:51:39          Honda_Jazz_1.3_DSi_i_VTEC_IMA_CVT_Comfort   
49976  2016-03-19 18:56:05  Audi_80_Avant_2.6_E__Vollausstattung!!_Einziga...   
49977  2016-03-31 18:37:18                        Mercedes_Benz_C200_Cdi_W203   
49978  2016-04-04 10:37:14                        Mercedes_Benz_E_200_Classic   
49979  2016-03-20 18:38:40                      Volkswagen_Polo_1.6_TDI_Style   
49980  2016-03-12 10:55:54                            Ford_Escort_Turnier_16V   
49981  2016-03-15 09:38:21             Opel_Astra_Kombi_mit_Anhaengerkupplung   
49982  2016-03-29 18:51:08            Skoda_Fabia_4_Tuerer_Bj:2004__85.000Tkm   
49983  2016-03-06 12:43:04                                      Ford_focus_99   
49984  2016-03-31 22:48:48  Student_sucht_ein__Anfaengerauto___ab_2000_BJ_...   
49985  2016-04-02 16:38:23                          Verkaufe_meinen_vw_vento!   
49986  2016-04-04 20:46:02  Chrysler_300C_3.0_CRD_DPF_Automatik_Voll_Ausst...   
49987  2016-03-22 20:47:27  Audi_A3_Limousine_

In [253]:
autos.info()
autos.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
dateCrawled            50000 non-null object
name                   50000 non-null object
seller                 50000 non-null object
offerType              50000 non-null object
price                  50000 non-null object
abtest                 50000 non-null object
vehicleType            44905 non-null object
yearOfRegistration     50000 non-null int64
gearbox                47320 non-null object
powerPS                50000 non-null int64
model                  47242 non-null object
odometer               50000 non-null object
monthOfRegistration    50000 non-null int64
fuelType               45518 non-null object
brand                  50000 non-null object
notRepairedDamage      40171 non-null object
dateCreated            50000 non-null object
nrOfPictures           50000 non-null int64
postalCode             50000 non-null int64
lastSeen               50000 non-null obj

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

#### Our dataset contains 50,000 entries in 20 rows, most of which are full on non-null values. The dtypes of the columns are mostly 'object' (strings) or int64. All columns seem to be populated with a single entry for each row, with spaces replaced by underscores. As expected, many columns are written in German. 

In [254]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [255]:
# map cleaned column names using camel case
mapped_col = {'dateCrawled': 'date_crawled',
              'name': 'name',
              'seller': 'seller',
             'offerType': 'offer_type',
             'price': 'price',
             'abtest': 'abtest',
             'vehicleType': 'vehicle_type',
             'yearOfRegistration': 'registration_year',
             'gearbox': 'gearbox',
             'powerPS': 'power_ps',
             'model': 'model',
             'odometer': 'odometer',
             'monthOfRegistration': 'registration_month',
             'fuelType': 'fuel_type',
             'brand': 'brand',
             'notRepairedDamage': 'unrepaired_damage',
             'dateCreated': 'ad_created',
             'nrOfPictures': 'nr_of_picutres',
             'poastalCode': 'postal_code',
             'lastSeen': 'last_seen'
             }

autos = autos.rename(mapped_col, axis='columns')

In [256]:
autos

date_crawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
5      2016-03-21 13:47:45  Chrysler_Grand_Voyager_2.8_CRD_Aut.Limited_Sto...   
6      2016-03-20 17:55:21  VW_Golf_III_GT_Special_Electronic_Green_Metall...   
7      2016-03-16 18:55:19                               Golf_IV_1.9_TDI_90PS   
8      2016-03-22 16:51:34                                         Seat_Arosa   
9      2016-03-16 13:47:02          Renault_Megane_Scenic_1.6e_RT_Klimaanlage   
10     2016-03-15 01:41:36                       VW_Golf_Tuning_in_siber/grau   
11     2016-03-16 18:45:34                         Mercedes_A140_Motorschaden   
12     2016-03-31 19:48:22  Smart_smart_fortwo_coupe_softouch_pure_MHD_Pan...   
13     2016-03-23 10:48:32                                 Audi_A3_1.6_tuning   
14     2016-03-23 11:50:46  Renault_Clio_3__Dynamique_1.2__16_V;_viele_Ver...   
15     2016-04-01 12:06:20        Corvette_C3_Coupe_T_Top_Crossfire_Injection   
16     2016-03-16 14:59:02                                Opel_Vectra_B_Kombi   
17     2016-03-29 11:46:22                          Volkswagen_Scirocco_2_G60   
18     2016-03-26 19:57:44               Verkaufen_mein_bmw_e36_320_i_touring   
19     2016-03-17 13:36:21        mazda_tribute_2.0_mit_gas_und_tuev_neu_2018   
20     2016-03-05 19:57:31  Audi_A4_Avant_1.9_TDI_*6_Gang*AHK*Klimatronik*...   
21     2016-03-06 19:07:10                      Porsche_911_Carrera_4S_Cabrio   
22     2016-03-28 20:50:54                               MINI_Cooper_S_Cabrio   
23     2016-03-10 19:55:34   Peugeot_Boxer_2_2_HDi_120_Ps_9_Sitzer_inkl_Klima   
24     2016-04-03 11:57:02                         BMW_535i_xDrive_Sport_Aut.   
25     2016-03-21 21:56:18  Ford_escort_kombi_an_bastler_mit_ghia_ausstattung   
26     2016-04-03 22:46:28                                Volkswagen_Polo_Fox   
27     2016-03-27 18:45:01             Hat_einer_Ahnung_mit_Ford_Galaxy_HILFE   
28     2016-03-19 21:56:19                                      MINI_Cooper_D   
29     2016-04-02 12:45:44  Mercedes_Benz_E_320_T_CDI_Avantgarde_DPF7_Sitz...   
...                    ...                                                ...   
49970  2016-03-21 22:47:37  c4_Grand_Picasso_mit_Automatik_Leder_Navi_Temp...   
49971  2016-03-29 14:54:12                                         W.Lupo_1.0   
49972  2016-03-26 22:25:23          Mercedes_Benz_Vito_115_CDI_Extralang_Aut.   
49973  2016-03-27 05:32:39                   Mercedes_Benz_SLK_200_Kompressor   
49974  2016-03-20 10:52:31  Golf_1_Cabrio_Tuev_Neu_viele_Extras_alles_eing...   
49975  2016-03-27 20:51:39          Honda_Jazz_1.3_DSi_i_VTEC_IMA_CVT_Comfort   
49976  2016-03-19 18:56:05  Audi_80_Avant_2.6_E__Vollausstattung!!_Einziga...   
49977  2016-03-31 18:37:18                        Mercedes_Benz_C200_Cdi_W203   
49978  2016-04-04 10:37:14                        Mercedes_Benz_E_200_Classic   
49979  2016-03-20 18:38:40                      Volkswagen_Polo_1.6_TDI_Style   
49980  2016-03-12 10:55:54                            Ford_Escort_Turnier_16V   
49981  2016-03-15 09:38:21             Opel_Astra_Kombi_mit_Anhaengerkupplung   
49982  2016-03-29 18:51:08            Skoda_Fabia_4_Tuerer_Bj:2004__85.000Tkm   
49983  2016-03-06 12:43:04                                      Ford_focus_99   
49984  2016-03-31 22:48:48  Student_sucht_ein__Anfaengerauto___ab_2000_BJ_...   
49985  2016-04-02 16:38:23                          Verkaufe_meinen_vw_vento!   
49986  2016-04-04 20:46:02  Chrysler_300C_3.0_CRD_DPF_Automatik_Voll_Ausst...   
49987  2016-03-22 20:47:27  Audi_A3_Limousine

#### I've updated each of the column names to be more intuitive to work with. For any ambiguous names, I've changed the name to more properly reflect the column contents. Further, each column was changed from camel case to snake case to avoid confusion. 

In [257]:
autos.describe(include='all')

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-03-22 09:51:06  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  registration_year  gearbox      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  nr_of_picutres    postalCode            last_seen  
count                 50000         50000.0  50000.000000                50000  
unique                   76             NaN           NaN                39481  
top     2016-04-03 00:00:00             NaN           NaN  2016-04-07 06:17:27  
freq                   1946             NaN           NaN                    8  
mean                    NaN             0.0  50813.627300                  NaN  
std                     NaN             0.0  25779.747957                  NaN  
min                     NaN             0.0   1067.000000                  NaN  
25%                     NaN             0.0  30451.000000                  NaN  
50%                     NaN             0.0  49577.000000                  NaN  
75%                     NaN             0.0  71540.000000                  NaN  
max                     NaN             0.0  99998.000000                  NaN

In [258]:
print(autos['fuel_type'].unique())
print()
print(autos['seller'].value_counts())
print()
print(autos['offer_type'].value_counts())
print()
print(autos['abtest'].value_counts())
print()
print(autos['gearbox'].value_counts())
print()
print(autos['unrepaired_damage'].value_counts())

['lpg' 'benzin' 'diesel' nan 'cng' 'hybrid' 'elektro' 'andere']

privat        49999
gewerblich        1
Name: seller, dtype: int64

Angebot    49999
Gesuch         1
Name: offer_type, dtype: int64

test       25756
control    24244
Name: abtest, dtype: int64

manuell      36993
automatik    10327
Name: gearbox, dtype: int64

nein    35232
ja       4939
Name: unrepaired_damage, dtype: int64


In [259]:
print(autos['price'].head())
print()
print(autos['date_crawled'].dtypes)
print()
print(autos['ad_created'].dtypes)
print()
print(autos['last_seen'].dtypes)
print()
print(autos['registration_year'].dtypes)

0    $5,000
1    $8,500
2    $8,990
3    $4,350
4    $1,350
Name: price, dtype: object

object

object

object

int64


#### Preliminary investigations show that both the 'seller' and 'offer_type' columns contain idential values for every row except 1, so we can eliminate those columns. The 'price' and 'odometer' columns will need to be convereted from an object to a float after removing the dollar sign and changing the column name to mention the price is in Euros and distance in kms. Each column containing dates will also need to be changed from object to a datetime. 

#### The 'registration_month' and 'power_ps' columns both have minimum values of 0, so those will need to be further investigated to make sure 0 values do not affect the analysis. Also, the 'registration_year' column has a min, max of 1000, 9999, which will need to be examined.

In [260]:
years = autos['registration_year'].value_counts()
print(years.sort_index())
print()
print(autos['registration_month'].value_counts().sort_index())
print()
print(autos['power_ps'].value_counts().sort_index())

1000       1
1001       1
1111       1
1500       1
1800       2
1910       9
1927       1
1929       1
1931       1
1934       2
1937       4
1938       1
1939       1
1941       2
1943       1
1948       1
1950       3
1951       2
1952       1
1953       1
1954       2
1955       2
1956       5
1957       2
1958       4
1959       7
1960      34
1961       6
1962       4
1963       9
        ... 
2001    2703
2002    2533
2003    2727
2004    2737
2005    3015
2006    2708
2007    2304
2008    2231
2009    2098
2010    1597
2011    1634
2012    1323
2013     806
2014     666
2015     399
2016    1316
2017    1453
2018     492
2019       3
2800       1
4100       1
4500       1
4800       1
5000       4
5911       1
6200       1
8888       1
9000       2
9996       1
9999       4
Name: registration_year, Length: 97, dtype: int64

0     5075
1     3282
2     3008
3     5071
4     4102
5     4107
6     4368
7     3949
8     3191
9     3389
10    3651
11    3360
12    3447
Name: registr

#### Further investigation reveals a handful of entries for cars registered before 1900 and after 2016 which are both impossible (data was scraped in 2016). Those entries will be removed. Additionally, 5000+ cars list the registration month as 0. 

#### The 'power_ps' column lists a range of values from 0-9999. As the average PS of a car is around 120, with the minimum being 1 or 2 for mopeds and the maximum known PS of a car being around 700, we'll need to remove any values outside of those bounds. It is also adviseable to investigate cars toward the lower or higher ends to see if their reported values make sense. 

In [261]:
# Clean and rename 'price' and 'odometer' columns
autos['price_euro'] = autos['price'].str.replace('$', '').str.replace(',', '').astype(float)
autos['odometer_km'] = autos['odometer'].str.replace(',', '').str.replace('km','').astype(float)
autos = autos.drop(['price', 'odometer'], axis=1)

In [262]:
# Clean 'registration_year' and 'registration_month' columns
# and drop bad rows
bad_reg_year = (autos['registration_year'] > 2016) | (autos['registration_year'] < 1900)
autos = autos.loc[~bad_reg_year]

bad_reg_months = autos['registration_month'] == 0
autos = autos.loc[~bad_reg_months]

# Clean 'power_ps' column and drop and rows below 1 or above 700
bad_power = (autos['power_ps'] > 700) | (autos['power_ps'] < 1)
autos = autos.loc[~bad_power]

# Drop 'seller' and 'offer_type' columns
autos = autos.drop(['seller', 'offer_type'], axis=1)


In [263]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 40253 entries, 0 to 49999
Data columns (total 18 columns):
date_crawled          40253 non-null object
name                  40253 non-null object
abtest                40253 non-null object
vehicle_type          39120 non-null object
registration_year     40253 non-null int64
gearbox               39732 non-null object
power_ps              40253 non-null int64
model                 38887 non-null object
registration_month    40253 non-null int64
fuel_type             38623 non-null object
brand                 40253 non-null object
unrepaired_damage     35313 non-null object
ad_created            40253 non-null object
nr_of_picutres        40253 non-null int64
postalCode            40253 non-null int64
last_seen             40253 non-null object
price_euro            40253 non-null float64
odometer_km           40253 non-null float64
dtypes: float64(2), int64(5), object(11)
memory usage: 5.8+ MB


In [264]:
# import datetime to convert date columns to dtypes datetime for
# date_crawled, ad_created, last_seen columns
import datetime

# Create function to convert values in col to datetime
def col_to_datetime(col):
    col_dt = pd.to_datetime(col, format='%Y-%m-%d %H:%M:%S')
    return col_dt

autos['date_crawled'] = col_to_datetime(autos['date_crawled'])
autos['ad_created'] = col_to_datetime(autos['ad_created'])
autos['last_seen'] = col_to_datetime(autos['last_seen'])

In [265]:
print(autos['date_crawled'].dtypes)

datetime64[ns]


In [266]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 40253 entries, 0 to 49999
Data columns (total 18 columns):
date_crawled          40253 non-null datetime64[ns]
name                  40253 non-null object
abtest                40253 non-null object
vehicle_type          39120 non-null object
registration_year     40253 non-null int64
gearbox               39732 non-null object
power_ps              40253 non-null int64
model                 38887 non-null object
registration_month    40253 non-null int64
fuel_type             38623 non-null object
brand                 40253 non-null object
unrepaired_damage     35313 non-null object
ad_created            40253 non-null datetime64[ns]
nr_of_picutres        40253 non-null int64
postalCode            40253 non-null int64
last_seen             40253 non-null datetime64[ns]
price_euro            40253 non-null float64
odometer_km           40253 non-null float64
dtypes: datetime64[ns](3), float64(2), int64(5), object(8)
memory usage: 5.8+ 

#### The dates have been changed to datetimes
#### The prices and odometer columns have been changd to ints and renamed
#### Rows with erroneous values for registration year, month, and power have been removed

In [267]:
print(autos['price_euro'].describe())
print()
print(autos['odometer_km'].describe())

count    4.025300e+04
mean     1.052974e+04
std      5.257858e+05
min      0.000000e+00
25%      1.400000e+03
50%      3.500000e+03
75%      8.199000e+03
max      1.000000e+08
Name: price_euro, dtype: float64

count     40253.000000
mean     124962.487268
std       39530.794031
min        5000.000000
25%      100000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64


#### Prices listed range from 0EUR to 100,000,000EUR, so we will need to investigate those and exclude any outliers.
#### The odometer readings rand from 5,000km to 150,000km which seems reasonable

In [268]:
print(autos['price_euro'].value_counts().sort_index(ascending=False).head(10))
print()
print(autos['price_euro'].value_counts().sort_index().head(10))

99999999.0    1
27322222.0    1
12345678.0    1
11111111.0    1
10000000.0    1
1234566.0     1
999999.0      2
999990.0      1
350000.0      1
345000.0      1
Name: price_euro, dtype: int64

0.0     616
1.0      62
3.0       1
10.0      2
11.0      1
13.0      1
20.0      1
30.0      1
45.0      1
47.0      1
Name: price_euro, dtype: int64


#### For our purposes we will exclude any cars listed as less than 100EUR or more than 350,000EUR. Cars sold for parts may very well be worth as little as 100EURs, Additionally, cars in the 100,000EUR range may be extremely rare or collectible vehicles, so the prices is not surprising. Resultingly, median may be a more accurate average measure for listed car price.

In [269]:
autos = autos.loc[autos['price_euro'].between(100, 350001)]

In [276]:
autos

date_crawled                                               name  \
0     2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1     2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2     2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3     2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
5     2016-03-21 13:47:45  Chrysler_Grand_Voyager_2.8_CRD_Aut.Limited_Sto...   
6     2016-03-20 17:55:21  VW_Golf_III_GT_Special_Electronic_Green_Metall...   
7     2016-03-16 18:55:19                               Golf_IV_1.9_TDI_90PS   
9     2016-03-16 13:47:02          Renault_Megane_Scenic_1.6e_RT_Klimaanlage   
12    2016-03-31 19:48:22  Smart_smart_fortwo_coupe_softouch_pure_MHD_Pan...   
13    2016-03-23 10:48:32                                 Audi_A3_1.6_tuning   
14    2016-03-23 11:50:46  Renault_Clio_3__Dynamique_1.2__16_V;_viele_Ver...   
15    2016-04-01 12:06:20        Corvette_C3_Coupe_T_Top_Crossfire_Injection   
16    2016-03-16 14:59:02                                Opel_Vectra_B_Kombi   
17    2016-03-29 11:46:22                          Volkswagen_Scirocco_2_G60   
19    2016-03-17 13:36:21        mazda_tribute_2.0_mit_gas_und_tuev_neu_2018   
20    2016-03-05 19:57:31  Audi_A4_Avant_1.9_TDI_*6_Gang*AHK*Klimatronik*...   
21    2016-03-06 19:07:10                      Porsche_911_Carrera_4S_Cabrio   
22    2016-03-28 20:50:54                               MINI_Cooper_S_Cabrio   
23    2016-03-10 19:55:34   Peugeot_Boxer_2_2_HDi_120_Ps_9_Sitzer_inkl_Klima   
24    2016-04-03 11:57:02                         BMW_535i_xDrive_Sport_Aut.   
26    2016-04-03 22:46:28                                Volkswagen_Polo_Fox   
28    2016-03-19 21:56:19                                      MINI_Cooper_D   
29    2016-04-02 12:45:44  Mercedes_Benz_E_320_T_CDI_Avantgarde_DPF7_Sitz...   
31    2016-03-14 16:53:09                          Noch_gut_erhaltenen_C_320   
33    2016-03-15 12:25:26  Audi_A3_2.0_TDI_Sportback_DPF_Kupplung_defekt_...   
34    2016-04-05 07:55:07                            Bmw_530d__navi_tuv03/18   
36    2016-03-16 23:46:31          Ford_Mondeo_2.0_Turnier_TDCi_DPF_Titanium   
37    2016-03-23 16:47:46                 Seat_Altea_XL_2.0_TDI_CR_DPF_Style   
38    2016-03-21 15:51:10                        Volkswagen_Golf_1.4_Special   
39    2016-03-11 13:58:38                 Mercedes_C_180_Kombi__zu_verkaufen   
...                   ...                                                ...   
49963 2016-03-26 19:52:59            Mercedes_Benz_B_200_CDI_Special_Edition   
49964 2016-03-10 16:46:43                                 Audi_2_7_TDI_AVANT   
49965 2016-03-11 10:53:51          Opel_Astra_1.6_Lenkradheizung~Sitzheizung   
49966 2016-04-02 19:49:19  Citroën_C1_1.0_**Euro4**TÜV_OKT_2017**Scheiten...   
49967 2016-03-12 11:50:17                      VW_Passat_2_0_TDI_comfortline   
49968 2016-04-01 17:49:15                  Mercedes_Benz_190_D_2.5_Automatik   
49969 2016-03-17 18:49:02             Nissan_X_Trail_2.2_dCi_4x4_Sport_m.AHZ   
49970 2016-03-21 22:47:37  c4_Grand_Picasso_mit_Automatik_Leder_Navi_Temp...   
49971 2016-03-29 14:54:12                                         W.Lupo_1.0   
49972 2016-03-26 22:25:23          Mercedes_Benz_Vito_115_CDI_Extralang_Aut.   
49973 2016-03-27 05:32:39                   Mercedes_Benz_SLK_200_Kompressor   
49975 2016-03-27 20:51:39          Honda_Jazz_1.3_DSi_i_VTEC_IMA_CVT_Comfort   
49976 2016-03-19 18:56:05  Audi_80_Avant_2.6_E__Vollausstattung!!_Einziga...   
49977 2016-03-31 18:37:18                        Mercedes_Benz_C200_Cdi_W203   
49978 2016-04-04 10:37:14                        Mercedes_Benz_E_200_Classic   
49979 2016-03-20 18:38:40                      Volkswagen_Polo_1.6_TDI_Style   
49980 2016-03-12 10:55:54                            Ford_Escort_Turnier_16V   
49981 2016-03-15 09:38:21             Opel_Astra_Kombi_mit_Anhaengerkupplung   
49983 2016-03

In [275]:
autos['brand'].value_counts().head(10)

volkswagen       8271
bmw              4552
opel             4050
mercedes_benz    3916
audi             3557
ford             2664
renault          1758
peugeot          1171
fiat              984
seat              736
Name: brand, dtype: int64

#### A majority of the cars listed are of 10 different brands, with Volkswagen dominating the group. Next we will determine the average prices of each brand of car for the top 10 brands in the dataset

In [277]:
top10_brands = autos['brand'].value_counts().head(10).index

In [309]:
# Create dictionary to hold median price values
top10_brands_median_price = {}

# Loop through the top 10 brands and, for each, find the median
# of the 'price_euro' column in autos. Add brand and price to dictionary
for brand in top10_brands:
    median = autos.loc[autos['brand'] == brand, 'price_euro'].median()
    top10_brands_median_price[brand] = median
    
# top10_brands_median_price = sorted(top10_brands_median_price.items(), key=lambda kv:kv[1], reverse=True)
top10_brands_median_price

{'audi': 7100.0,
 'bmw': 6500.0,
 'fiat': 1950.0,
 'ford': 2199.0,
 'mercedes_benz': 5794.5,
 'opel': 1990.0,
 'peugeot': 2400.0,
 'renault': 1550.0,
 'seat': 3333.0,
 'volkswagen': 3549.0}

#### As noted before, due to the possible presence of extremely rare/collectors cars, we are using the median price here as a measure of average. 

####  Audi tops the list with an average price of 7.100,00EUR, closely followed by BMW. 

#### Renault cars average about one-quarter of the price of Audis listed at 1.550,00EUR

#### While Volkswagen is by far the most common car brand in the data set, it is, on average, only the 4th most expensive car of the top 10 brands listed, with an average price of 3.549,00EUR.

In [323]:
# Calculate the mean mileage for each car of the top 10 brands
top10_mean_mileage = {}
for brand in top10_brands:
    mean_mileage = autos.loc[autos['brand'] == brand, 'odometer_km'].mean()
    top10_mean_mileage[brand] = mean_mileage

# Convert dictionaries to pd.Series    
bmp_series = pd.Series(top10_brands_median_price)    
bmm_series = pd.Series(top10_mean_mileage)

# Create a pd.DataFrame from first series and then add the other
top10_price_miles = pd.DataFrame(bmm_series, columns=['mean_mileage'])
top10_price_miles['median_price'] = bmp_series
top10_price_miles.round()


mean_mileage  median_price
audi               127924.0        7100.0
bmw                132202.0        6500.0
fiat               116067.0        1950.0
ford               123716.0        2199.0
mercedes_benz      130462.0        5794.0
opel               128332.0        1990.0
peugeot            125312.0        2400.0
renault            126610.0        1550.0
seat               120265.0        3333.0
volkswagen         127680.0        3549.0

#### With the average mileage and price next to each other in our dataframe, we can see that there doesn't appear to be a correlation between the two. The range of average mileages is very narrow and such small variation should not be expected to explain the large price differential

## Further analysis could center on comparing prices of cars of the same make and model based on the presence of unrepaired damage, prices based on mileage, frequencies of 